### PR TITLE
Resource get texture info

### LINKED ISSUE #7564: Functionality to get texture size of atlases
Is it possible to add some functionality to get texture size of atlas?

The problem is: Sometimes when working with shaders it is necessary to know the exact size of the atlas texture for calculations. For example, to draw outline of the same thickness for sprites located in different atlases. As long as one atlas is used, there are no problems, but if there are several atlases, and they are of different sizes (for example, 2048 * 2048 and 1024 * 1024), the algorithms will draw outline of different thicknesses.

### PR DESCRIPTION
**Pull request by Jhonnyg**
Added the new function `resource.get_texture_info` to get metadata from a texture resource or handle:

```
local my_texture_info = resource.get_texture_info(my_texture_path)

-- my_texture_info is a table that contains
{
     handle, -- the numeric handle of the texture
     width, -- width of the texture
     height, -- height of the texture
     depth, -- depth of the texture (1 for a 2D texture, 6 for a cube map and [1...slice-count] for an array texture)
     mipmaps, -- number of mipmaps the texture has
     type -- texture type
}
```

Fixes #7564
Fixes #7362

## PR checklist

* [x] Code
	* [x] Add engine and/or editor unit tests.
	* [x] New and changed code follows the overall code style of existing code
	* [x] Add comments where needed
* [x] Documentation
	* [x] Make sure that API documentation is updated in code comments
	* [ ] Make sure that manuals are updated (in github.com/defold/doc)
* [x] Prepare pull request and affected issue for automatic release notes generator
	* [x] Pull request - Write a message that explains what this pull request does. What was the problem? How was it solved? What are the changes to APIs or the new APIs introduced? This message will be used in the generated release notes. Make sure it is well written and understandable for a user of Defold.
	* [x] Pull request - Write a pull request title that in a sentence summarises what the pull request does. Do not include "Issue-1234 ..." in the title. This text will be used in the generated release notes.
	* [x] Pull request - Link the pull request to the issue(s) it is closing. Use on of the [approved closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
	* [x] Affected issue - Assign the issue to a project. Do not assign the pull request to a project if there is an issue which the pull request closes.
	* [ ] Affected issue - Assign the "breaking change" label to the issue if introducing a breaking change.
	* [ ] Affected issue - Assign the "skip release notes" is the issue should not be included in the generated release notes.

### CODE REVIEW

File: engine/gamesys/src/gamesys/scripts/script_resource.cpp [hunk 1]
```diff
@@ -2573,6 +2718,7 @@ static void LuaInit(lua_State* L)
 
     SETGRAPHICSCONSTANT(TEXTURE_TYPE_2D);
     SETGRAPHICSCONSTANT(TEXTURE_TYPE_CUBE_MAP);
+    SETGRAPHICSCONSTANT(TEXTURE_TYPE_2D_ARRAY);
```
Comment (Jhonnyg) [inline engine/gamesys/src/gamesys/scripts/script_resource.cpp:2721]: Added because even if we don't support creating dynamic array textures (yet), we can still get the info from a texture resource that has been created as an array texture!

File: engine/graphics/src/opengl/graphics_opengl.cpp [hunk 2]
```diff
@@ -500,25 +500,6 @@ static void LogFrameBufferError(GLenum status)
         return GL_FALSE;
     }
 
-    static Type ShaderDataTypeToGraphicsType(ShaderDesc::ShaderDataType shader_type)
```
Comment (Jhonnyg) [inline engine/graphics/src/opengl/graphics_opengl.cpp:503]: unused

File: engine/graphics/src/vulkan/graphics_vulkan.cpp [hunk 3]
```diff
@@ -3635,6 +3644,27 @@ namespace dmGraphics
...
+        if (type == ASSET_TYPE_TEXTURE)
+        {
+            return GetAssetFromContainer<Texture>(context->m_AssetHandleContainer, asset_handle) != 0;
+        }
+        else if (type == ASSET_TYPE_RENDER_TARGET)
+        {
+            return GetAssetFromContainer<RenderTarget>(context->m_AssetHandleContainer, asset_handle) != 0;
+        }
+        return false;
+    }
```
Comment (Jhonnyg) [inline engine/graphics/src/vulkan/graphics_vulkan.cpp:3666]: we should generalize this to just take a reference to the container, all the adapters will do this identically

File: engine/gamesys/src/gamesys/scripts/script_resource.cpp [hunk 4]
```diff
@@ -714,11 +715,17 @@ static int CreateTexture(lua_State* L)
...
+    uint32_t width               = (uint32_t) CheckTableInteger(L, 2, "width");
+    uint32_t height              = (uint32_t) CheckTableInteger(L, 2, "height");
+    uint32_t format              = (uint32_t) CheckTableInteger(L, 2, "format");
+    uint32_t max_mipmaps         = (uint32_t) CheckTableInteger(L, 2, "max_mipmaps", 0);
+
+    // TODO: Texture arrays
+    if (!(type == dmGraphics::TEXTURE_TYPE_2D || type == dmGraphics::TEXTURE_TYPE_CUBE_MAP))
+    {
+        return luaL_error(L, "Unable to create texture, unsupported texture type '%s'.", dmGraphics::GetTextureTypeLiteral(type));
+    }
```
Comment (Jhonnyg) [inline engine/gamesys/src/gamesys/scripts/script_resource.cpp:728]: I want to do this in a separate PR, so just drop an error for now..

File: engine/graphics/src/graphics.h [hunk 5]
```diff
@@ -216,33 +219,34 @@ namespace dmGraphics
...
-        const void* m_Data;
-        uint32_t m_DataSize;
-        uint16_t m_MipMap;
-        uint16_t m_Width;
-        uint16_t m_Height;
-        uint16_t m_Depth; // For array texture, this is slice count
+        TextureWrap   m_UWrap;
+        TextureWrap   m_VWrap;
+        const void*   m_Data;
+        uint32_t      m_DataSize;
```
Comment (Jhonnyg) [inline engine/graphics/src/graphics.h:238]: Sorted on size

File: engine/render/src/render/render_script.cpp [hunk 6]
```diff
@@ -1080,7 +1080,7 @@ namespace dmRender
             luaL_error(L, "Unable to create render target.");
         }
 
-        lua_pushinteger(L, render_target);
+        lua_pushnumber(L, render_target);
```
Comment (Jhonnyg) [inline engine/render/src/render/render_script.cpp:1083]: win32 fix, integer is pushed as 32b and number as 64

File: engine/gamesys/src/gamesys/scripts/script_resource.cpp [hunk 7]
```diff
@@ -1003,6 +1016,131 @@ static int SetTexture(lua_State* L)
     return 0;
 }
 
+/*# get texture info
+ * Gets texture info from a texture resource path or a texture handle
+ *
+ * @name resource.get_texture_info
+ *
+ * @param path [type:hash|string|handle] The path to the resource or a texture handle
```
Comment (Jhonnyg) [inline engine/gamesys/src/gamesys/scripts/script_resource.cpp:1024]: Note to reviewer: the function supports both a path and a handle because I think it will make sense to query handles created from elsewhere, but now that I think about it maybe it should be in a different namespace?

File: engine/gamesys/src/gamesys/scripts/script_resource.cpp [hunk 8]
```diff
@@ -1003,6 +1016,131 @@ static int SetTexture(lua_State* L)
...
+    uint32_t texture_width               = dmGraphics::GetTextureWidth(texture_handle);
+    uint32_t texture_height              = dmGraphics::GetTextureHeight(texture_handle);
+    uint32_t texture_depth               = dmGraphics::GetTextureDepth(texture_handle);
+    uint32_t texture_mipmaps             = dmGraphics::GetTextureMipmapCount(texture_handle);
+    dmGraphics::TextureType texture_type = dmGraphics::GetTextureType(texture_handle);
+
+    lua_newtable(L);
+
+    lua_pushnumber(L, texture_handle);
+    lua_setfield(L, -2, "handle");
```
Comment (JCash) [inline engine/gamesys/src/gamesys/scripts/script_resource.cpp:1123]: Is this really safe for all possible 64 bit signed integers?
Some of those bit patterns will represent NaN's etc, are we guaranteed to get the same number back? And does it work the same way for both Lua 5.1 and LuaJIT?

I'm wondering if using it as a light user data instead removes those questions altogether:
https://www.lua.org/manual/5.1/manual.html#lua_pushlightuserdata

File: engine/graphics/src/graphics.h [hunk 9]
```diff
@@ -216,33 +219,34 @@ namespace dmGraphics
...
-        bool m_SubUpdate;
         uint32_t m_X;
         uint32_t m_Y;
         uint32_t m_Z; // For array texture, this is the slice to set
+
+        uint16_t m_Width;
+        uint16_t m_Height;
+        uint16_t m_Depth; // For array texture, this is slice count
+        uint8_t  m_MipMap    : 7;
+        uint8_t  m_SubUpdate : 1;
```
Comment (JCash) [inline engine/graphics/src/graphics.h:249]: What's the estimated upper limit of the depth? 255?
Then it could be:
```
uint16_t m_Depth : 8;
uint16_t m_MipMap : 7;
uint16_t m_SubUpdate : 1;
```

File: engine/graphics/src/graphics.h (hunk 5 above)
Comment (JCash) [reply engine/graphics/src/graphics.h:238]: Not all the way though.
The pointer is potentially larger than the enums.

File: engine/render/src/render/render_script.cpp (hunk 6 above)
Comment (JCash) [reply engine/render/src/render/render_script.cpp:1083]: Same concern here then (re: light user data)

**Review by JCash — CHANGES_REQUESTED** [review]
I'm a bit concerned with using our large handles as doubles.
Pleasy clarify (with a comment in code) that this is safe, or possibly use light user data

File: engine/graphics/src/graphics.h (hunk 9 above)
Comment (Jhonnyg) [reply engine/graphics/src/graphics.h:249]: The upper depth limit is dictated by the graphics card vendors for 3D/array textures, I don't think we have an absolute max for depth slices in our engine. We loosely support 16k textures so the depth should also adhere to that limit

File: engine/gamesys/src/gamesys/scripts/script_resource.cpp (hunk 8 above)
Comment (Jhonnyg) [reply engine/gamesys/src/gamesys/scripts/script_resource.cpp:1123]: Hmm, that's a good point. The reason I opted for the 64b handle was to store the type information in the handle itself so we know what it's supposed to represent. I.e if you create a texture via the resource namespace and pass it to another system (render script most likely) and call render.enable_texture with the handle we need to be able to distinguish between a render target and a texture. Alternatively we can represent the handle as a 32b value and store the type information in the graphics adapters, but it's a less 'neat' solution since we need another indirection just to store that extra bit (or maybe bits in the future). Maybe I'm wrong but with lightuserdata I'm not sure we solve this issue anyway since we don't serialize that type in the message system right? And they are 32b on win32 so I guess the only thing it helps with in that case is to preserve the bits in case lua does something strange with integers

File: engine/graphics/src/graphics.h (hunk 5 above)
Comment (Jhonnyg) [reply engine/graphics/src/graphics.h:238]: Fixed

**Comment by Jhonnyg** [comment]
After some discussions with @JCash , we concluded that we can use doubles for the handles so we should be all good!
